### ISSUE TITLE
Parametric size support isn't well-defined

### ISSUE THREAD

**Issue by kaushikcfd** [inducer/pytato#310]
For some computation graphs, the shape is dependent on value of the computation graph's parameters. One example of such a computation graph is:
```python
n = pt.make_size_param("n")
a = pt.make_placeholder("a", shape=(10,n), dtype="float64")
out = pt.squeeze(a)
```

Notice how the shape of `out` depends on the value of "n" passed to the program. However, we do not account for such shape specialization based on the values of the parameters.

**Comment by inducer** [comment]
I agree that there's an issue here, but, at least for the example you show, there are a number of ways out:
- We could simply ban `squeeze` for parametrically-shaped arrays.
- We could define `squeeze` to assume that parametric shape components are non-1.
- We could add (likely polyhedral) assumption tracking on shape components.

There are other operations that are potentially problematic, such as broadcasting, which behaves differently based on whether shape components are one.

**Comment by kaushikcfd** [comment]
> We could add (likely polyhedral) assumption tracking on shape components.

Yikes, had forgotten about the crucial broadcasting operations. I guess we have dug ourselves into the problems as outlined in https://dl.acm.org/doi/pdf/10.1145/570440.570499, https://dl.acm.org/doi/pdf/10.1145/390009.804442. :)

**Comment by inducer** [comment]
Thanks for sharing those, certainly a blast from the past!

Not sure I totally agree though. While an APL compiler can't help but take the language definition literally, I feel that we have a greater amount of control in how we define operations on parametrically-shaped data, and we're IMO well within our right to make decisions that enable us to make decisions more statically. Yes, that means the behavior won't be strictly numpy if a parametric-length axis dynamically turns out to be 1-long. But I am skeptical that it is a good use of time and resources to address these problems. Put differnently, can you point to a workload where this would be a crucial enabler? (I can't. Most numpy code I know makes very static assumptions about the number of axes.)

**Comment by kaushikcfd** [comment]
> But I am skeptical that it is a good use of time and resources to address these problems. Put differnently, can you point to a workload where this would be a crucial enabler

I agree, can't think of any workloads where such value-based shape specialization plays a big role, but I wanted us to be well-defined. I guess an easy fix for all these would be documenting the requirement that "n > 1" for all size parameters in our graph.

**Comment by inducer** [comment]
> documenting the requirement that "n > 1" for all size parameters in our graph.

Maybe. It's easy to argue numpy compatibility subject to that requirement. It's also a bit restrictive.
